rfcs/20211207-softmax-v2/README.md

```diff
@@ -0,0 +1,144 @@
+# RFC: Softmax v2
+
+## Introduction
+
+Several primitives in the library were designed with operation descriptors containing only `data_desc` member, which covers source and destination memory descriptors.
+There was an assumption that such operations do not change neither memory format layout through execution nor require a different output data type. And would never be.
+Those primitives are batch and layer normalizations, eltwise, prelu, lrn, shuffle and softmax.
+
+These operations, except shuffle, assume that only floating-point types will be used on input and output of same data type will be produced.
+It is quite essential due to the nature and service of these operations, especially, normalizations.
+Also, outputs of these operations are clearly distinguished only in floating-point range of values.
+
+Unfortunately, that assumption did not consider int8 flows when computations may contain mixture of quantized data and floating-point data.
+Mentioned primitives are still executed in floating-point data types such as `f32`, `bf16` or `f16`, but it usually happens that followed by primitive would use quantized output from mentioned primitives.
+It means that every time such situation happens, a reorder from floating-point type to integer type appears.
+
+In Transformer-like models, including BERT, softmax layer appears in every Encoder and Decoder block.
+The number of blocks may vary but we may consider a base case when 12 of each are present resulting in 24 reorders bringing their overhead.
+
+Reorder can easily be incorporated into softmax operation, or any other operation, by letting it store the answer in requested data type.
+The problem that `data_desc` does not allow to do that and the only possible solution is to extend softmax descriptor to accept additional memory descriptor.
+Quantized output of mentioned primitives would not be meaningful in terms of operation essence, but this is what's used in production for inference and was tested accordingly from accuracy point of view.
+
+### Motivation
+
+Major motivation is free performance improvements for Transformer-like models for int8 inference.
+Additional motivation is to eliminate `logsoftmax` primitive API which was a result of preserving ABI for users back in the days.
```

Comment:
"back in the days" should definitely be "back in the day" but a less informal term might be better, maybe "in earlier software versions."

Reply:
I like it, applied, thanks.